A different approach to the slider update problem

Dav1dK opened the pull request:
Let the tank tabs register the listener and have a attribute which
tracks wether the model update is because of the slider, to ignore it.
This means that the model recieves updates all the time, except when the
updates are by itself. This is different than before, because before the
window had only recieved updates, if the control elements were disabled.

Additionally, improve some other small things.

Dav1dK committed:
[simulation] Remove setOverflowClosedHandler()

setResetHandler can be used for this purpose.

Dav1dK committed:
[simulation/view] Use textformatter to force integer input

Dav1dK committed:
[simulation/controller] Get rid of SimulationControlWindow.update()

Dav1dK committed:
[core/model/base] Make ProductionSite observerable

Dav1dK committed:
[simulation] Handle reset state of production site in controller

There is no need to have this in ProductionSite, as every reset
goes through the SimulationController

Dav1dK committed:
[simulation/view] A different solution to the update problem

Let the tank tabs register the listener and have a attribute which
tracks wether the model update is because of the slider, to ignore it.
This means that the model recieves updates all the time, except when the
updates are by itself. This is different than before, because before the
window had only recieved updates, if the control elements were disabled.

ByteHamster commented on src/osip-simulation-view/src/main/java/edu/kit/pse/osip/simulation/view/control/AbstractTankTab.java:
```diff
@@ -132,8 +121,19 @@ protected GridPane getGridPane() {
+     * Sets the listener that is notified of changes to valve thresholds.
+     * @param listener The Consumer that gets all changes to valve thresholds.
+     */
+    void setValveListener(BiConsumer<Pipe, Byte> listener) {
```
Würde protected schreiben

Dav1dK replied:
Intellij schlägt hier (und bei den anderen Kommentaren) Package Private vor, was ja auch Sinn macht, da es noch eingeschränkter ist. Soll ich es trotzdem ändern?

ByteHamster replied:
Ich weiß... Sieht finde ich trotzdem immer aus, als hätte man es vergessen :D

Dav1dK replied:
Na gut, von mir aus..

ByteHamster commented on src/osip-simulation-view/src/main/java/edu/kit/pse/osip/simulation/view/control/MixTankTab.java:
```diff
@@ -98,30 +105,32 @@ public void setControlsDisabled(boolean isDisable) {
+     * Sets the listener that is notified of changes in motor speed.
+     * @param listener The Consumer that gets all to the motorSpeed
+     */
+    void setMotorListener(Consumer<Integer> listener) {
```
Protected?

ByteHamster commented on src/osip-simulation-view/src/main/java/edu/kit/pse/osip/simulation/view/control/TankTab.java:
```diff
@@ -159,44 +166,44 @@ public void setControlsDisabled(boolean isDisable) {
     }
 
+    @Override
+    void setValveListener(BiConsumer<Pipe, Byte> listener) {
```
Protected

ByteHamster commented on src/osip-simulation-view/src/main/java/edu/kit/pse/osip/simulation/view/control/TankTab.java:
```diff
@@ -159,44 +166,44 @@ public void setControlsDisabled(boolean isDisable) {
+     * Sets the listener that is notified of changes in the temperature.
+     * @param listener The Consumer that gets all changes to Tank temperatures
+     */
+    void setTemperatureListener(BiConsumer<TankSelector, Float> listener) {
```
Protected

ByteHamster reviewed (commented):
Funktioniert bei mir

NikoWil reviewed (changes requested):
Funktioniert bei mir.
Allerdings ist mir noch ein Fehler aufgefallen (der auch vorher offenbar schon da war): wird ein Szenario gestartet und kommt es innerhalb oder nach dem Szenario zu einem Ueberlauf werden die Steuerelemente nicht geblockt und es kann wieder zum oeffnen sehr vieler OverFlowDialoge fuehren.

Ich suche noch nach einem Fix und wuerde den dann auch hier mit einbauen.

Problem gefunden:
Ausgangssituation: im SimulationController ist currentScenario != null
showOverflow wird aufgerufen
in der Methode in Zeile 223 ist die Bedingung wahr
Zeile 225 ruft currentSimulationView.scenarioFinished() auf

SimulationMainWindow.scenarioFinished() ruft in Zeile 252 menu.setScenarioFinished() auf

SimulationMenu.setScenarioFinished() aktiviert in Zeile 169 die Kontrollen

Moegliche Fixes:
- Wir koennen unterscheiden zwischen ScenarioFinished und ScenarioAborted und fuegen fuer letzteres neue Methoden ein, um es sauber zu trennen. Verursacht aber deutlich Overhead.
- Im SimulationController showOverflow aendern wir die Reihenfolge von <Kontrollen deaktivieren, Dialog anzeigen, Szenario beenden> zu <Szenario beenden, Kontrollen deaktivieren, Dialog anzeigen>. Dann besteht zwar die Gefahr, dass die Kontrollen fuer einen winzigen Moment aktiviert sind, allerdings ist zu dem Zeitpunkt garantiert noch kein Dialog in der Anzeige, sollte also gefahrlos sein.

Dav1dK commented:
Warum lassen wir nicht die Zeile 169 im SimulationMenu weg, die die Kontrollen aktiviert? Ich finde das gehört dort sowieso nicht hin. Diese Methode sollte sich nur um das Menu kümmern, schließlich ist es ja in der Datei SimulationMenu.java. Wenn die Kontrollen aktiviert werden müssen, dann sollte das doch der SimulationController machen können, oder?

ByteHamster commented:
Hatte mich auch ein bisschen gewundert, dass man das control window mit ins Menü schleppen muss

Dav1dK committed:
[simulation] Don't set controls enabled in scenarioFinished()

The menu should have nothing to do with the controls.
Also don't run the scenario finished handler if the scenario is manuelly
cancled, because this would enable the controls which should no happen
if there is an overflow.

Dav1dK commented:
Ich habe das mal implementiert.

Dav1dK committed:
[simulation/view] Make package private methods protected

As requested in #306.

NikoWil committed:
Merge branch 'master' into sliderfix
